File: algorithms/cpp/001-quicksort.cpp
#include <algorithm>
#include <array>
#include <iostream>
#include <random>
#include <span>
#include <tuple>
#include <vector>

std::tuple<std::span<int>, std::span<int>> part(std::span<int> &a) {
  int pivot = a.size() - 1;
  int j = 0;

  for (int i = 0; i < a.size(); i++) {
    if (a[i] < a[pivot]) {
      std::swap(a[i], a[j]);
      j++;
    }
  }

  std::swap(a[j], a[pivot]);
  std::span<int> n1{a.data(), a.data() + j};
  std::span<int> n2{a.data() + j + 1, a.data() + a.size()};

  return std::make_tuple(n1, n2);
}

void quicksort(std::span<int> &a) {
  if (a.size() > 1) {
    auto [p1, p2] = part(a);
    quicksort(p1);
    quicksort(p2);
  }
}

int main(void) {
  std::vector<int> a = {};
  std::random_device rd;
  std::mt19937 gen(rd());
  std::uniform_int_distribution<int> distrib_size(10, 50);
  std::uniform_int_distribution<int> distrib_elements(1, 100);

  auto print_a = [&]() {
    std::cout << "[";
    for (int i = 0; i < a.size(); i++) {
      std::cout << a[i] << (i + 1 < a.size() ? ", " : "");
    }
    std::cout << "]\n";
  };

  auto generate_a = [&]() {
    int n = distrib_size(gen);
    for (int i = 0; i < n; i++) {
      a.push_back(distrib_elements(gen));
    }
  };

  generate_a();
  print_a();
  std::span<int> s{a.data(), a.size()};
  quicksort(s);
  print_a();

  return 0;
}

File: algorithms/typescript/001-quicksort.ts
const part = (a: number[]): [number[], number[]] => {
  const pivot = a.length - 1;
  let j = 0;

  for (let i = 0; i < a.length; i++) {
    if (a[i] < a[pivot]) {
      [a[i], a[j]] = [a[j], a[i]];
      j++;
    }
  }

  [a[j], a[pivot]] = [a[pivot], a[j]];

  return [a.splice(0, j), a.splice(1)];
};

const quicksort = (a: number[]): number[] => {
  if (a.length > 1) {
    let [p1, p2] = part(a);

    let r1 = quicksort(p1);
    let r2 = quicksort(p2);

    let r = [...r1, a[0], ...r2];
    return r;
  }
  return a;
};

const main = () => {
  let a = [];

  const n = Math.floor(Math.random() * 41 + 10);
  for (let i = 0; i < n; i++) {
    a.push(Math.floor(Math.random() * 101 + 1));
  }

  console.log(a);
  a = quicksort(a);
  console.log(a);
};

main();
